### DISCUSSION TITLE
Compiler error: overflow evaluating the requirement

### DISCUSSION THREAD

**Discussion by ColinTimBarndt** [diesel-rs/diesel#2892] [Q&A] [answered]
Hi! I wanted to use the `diesel::sql_query` function with some bound values when I ran into this compile error:

```
error[E0275]: overflow evaluating the requirement `<diesel::query_builder::sql_query::UncheckedBind<diesel::query_builder::sql_query::UncheckedBind<diesel::query_builder::sql_query::UncheckedBind<SqlQuery, &str, diesel::sql_types::Text>, &[u8], diesel::sql_types::Binary>, &[u8], diesel::sql_types::Binary> as LimitDsl>::Output == _`

  --> authserver/src/database.rs:52:6
   |
52 |     .first::<models::InsertedUserId>(connection());
   |      ^^^^^
```

I don't know why I get this error and I already searched a lot about this error message and I'm not sure whether this is a compiler error or my fault, hoping for the latter. This is the code with the error:

```rs
let user_id = diesel::sql_query(
    "INSERT INTO users (username, password, salt) VALUES (?, ?, ?) RETURNING id",
)
.bind::<VarChar, &str>(username)
.bind::<Blob, &[u8]>(&hpassword)
.bind::<Blob, &[u8]>(&salt)
.first::<models::InsertedUserId>(connection());
```

The `connection()` function returns a `&'static SqliteConnection`. I am using rustc 1.57.0-nightly (e30b68353 2021-09-05).

<details><summary><b>Related definitions</b></summary>

### Database format (`up.sql`)

```sql
CREATE TABLE IF NOT EXISTS users (
	id UNSIGNED INTEGER PRIMARY KEY,
	username VARCHAR(32) NOT NULL,
	password BLOB NOT NULL,
	salt BLOB NOT NULL,
	created UNSIGNED INTEGER DEFAULT CURRENT_TIMESTAMP,
	refresh_id UNSIGNED CHAR(16) DEFAULT NULL
)
```

### Schema

```rs
table! {
	users (id) {
		id -> Integer,
		username -> VarChar,
		password -> Binary,
		salt -> Binary,
		created -> Timestamp,
		refresh_id -> Nullable<Text>,
	}
}
```

### InsertedUserId

```rs
#[derive(Debug, QueryableByName, Queryable)]
#[table_name="users"]
pub struct InsertedUserId {
	pub id: i32,
}
```

</details>

Could someone please help me? I have no clue what's going on here.

**Comment by weiznich** [comment, marked as answer]
Thats because `first` is not a valid method in this context. The required trait bounds for this method are not fulfilled. Use `get_result` instead.
The error message is really unfortunate. It would be great if you could fill a issue in the rustc repo for a bad diagnostic in this case. I believe the compiler should be able to point out the missing trait implementation instead emitting this overflow error.

**Reply by ColinTimBarndt** [reply]
Thank you so much, I think the main enemy here was rust-analyzer not knowing what code completions to give me here... `get_result` made my code compile without any errors though I could swear that I tried this before 🤷🏼

**Reply by weiznich** [reply]
If `first` showed up as code completion by rust-analyzer it might be worth to fill an issue there as well. Feel free to ping me from the corresponding issues

**Reply by ColinTimBarndt** [reply]
The problem was not that rust-analyzer gave me the wrong autocompletions, it gave me none proviously and just showed me `{Unknown}` or something similar as a type. For some reason after switching to stable Rust and back, activating and deactivating the Extension and restarting VS Code as a whole it now gives me autocompletions without highlighting errors. It's very weird.

I sadly don't know how to reproduce this making it hard to fill a issue
